File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;
class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Stack<Node> stack = new Stack<>();
        for(int i = 0 ; i < numbers.length ; i++){
            while(!stack.isEmpty() && stack.peek().y < numbers[i]){
                Node node = stack.pop();
                answer[node.x] = numbers[i];
            }
            stack.add(new Node(i, numbers[i]));
        }
        while(!stack.isEmpty()){
            Node node = stack.pop();
            answer[node.x] = -1;
        }
        return answer;
    }
    
    public static class Node{
        int x;
        int y;

        public Node(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.kt
import java.util.*

class Solution {
    fun solution(numbers: IntArray): IntArray {
        var answer: IntArray = IntArray(numbers.size){-1}
        var st = Stack<Pair<Int,Int>>()

        for(i in numbers.indices) {
            val num = numbers[i]
            while(st.isNotEmpty() && st.peek().second < num){
                val p = st.pop()
                answer[p.first] = num
            }
            st.add(Pair(i,num))
        }
        return answer
    }
}